plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/repository/M2RepositoryService.java

```diff
@@ -42,22 +50,100 @@
     } catch (CoreException ex) {
       throw new LibraryRepositoryServiceException(NLS.bind(Messages.ResolveArtifactError, mavenCoordinates), ex);
     }
   }
 
+  private static IClasspathAttribute[] getClasspathAttributes(LibraryFile libraryFile, Artifact artifact)
+                                                                              throws LibraryRepositoryServiceException {
+    try {
+      List<IClasspathAttribute> attributes = getAttributesForMavenCoordinates(artifact,
+                                                                              libraryFile.getMavenCoordinates());
+      if (libraryFile.isExport()) {
+        attributes.add(UpdateClasspathAttributeUtil.createDependencyAttribute(true /* isWebApp */));
+      } else {
+        attributes.add(UpdateClasspathAttributeUtil.createNonDependencyAttribute());
+      }
+      return attributes.toArray(new IClasspathAttribute[0]);
+    } catch (CoreException ex) {
+      throw new LibraryRepositoryServiceException("Could not create classpath attributes", ex);
+    }
+  }
+
+  private static List<IClasspathAttribute> getAttributesForMavenCoordinates(Artifact artifact,
+                                                                            MavenCoordinates mavenCoordinates) {
+    List<IClasspathAttribute> attributes = Lists.newArrayList(
+        JavaCore.newClasspathAttribute(CLASSPATH_ATTRIBUTE_REPOSITORY, mavenCoordinates.getRepository()),
+        JavaCore.newClasspathAttribute(CLASSPATH_ATTRIBUTE_GROUP_ID, artifact.getGroupId()),
+        JavaCore.newClasspathAttribute(CLASSPATH_ATTRIBUTE_ARTIFACT_ID, artifact.getArtifactId()),
+        JavaCore.newClasspathAttribute(CLASSPATH_ATTRIBUTE_TYPE, artifact.getType()),
+        JavaCore.newClasspathAttribute(CLASSPATH_ATTRIBUTE_VERSION, artifact.getVersion())
+        );
+    if (artifact.getClassifier() != null) {
+      attributes.add(JavaCore.newClasspathAttribute(CLASSPATH_ATTRIBUTE_CLASSIFIER, artifact.getClassifier()));
+    }
+    return attributes;
+  }
+
+  private IPath getSourceLocation(LibraryFile libraryFile) {
+    if (libraryFile.getSourceUri() == null) {
+      return getSourceJarLocation(libraryFile.getMavenCoordinates());
+    } else {
+      // download the file and return path to it
+      // TODO https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/800
+      return new Path("/downloaded/source/file"); //$NON-NLS-1$
+    }
+  }
+
+  private static IAccessRule[] getAccessRules(List<Filter> filters) {
+    IAccessRule[] accessRules = new IAccessRule[filters.size()];
+    int idx = 0;
+    for (Filter filter : filters) {
+      if (filter.isExclude()) {
+        IAccessRule accessRule = JavaCore.newAccessRule(new Path(filter.getPattern()), IAccessRule.K_NON_ACCESSIBLE);
+        accessRules[idx++] = accessRule;
+      } else {
+        IAccessRule accessRule = JavaCore.newAccessRule(new Path(filter.getPattern()), IAccessRule.K_ACCESSIBLE);
+        accessRules[idx++] = accessRule;
```

Comment:
You could do remove the if statement and code duplication by

```java
... rule = filter.isExcluse() ? IAccessRule.K_NON_ACCESSIBLE : IAccessRule.K_ACCESSIBLE;
accessRules[idx++] = JavaCore.newAccessRule(new Path(filter.getPattern()), rule);
```

but I'd say it's your preference.